File: src/com/castlefrog/agl/domains/connect4/Connect4Action.java
package com.castlefrog.agl.domains.connect4;

import java.util.ArrayList;
import java.util.List;

import com.castlefrog.agl.Action;

public final class Connect4Action implements Action {
    /** Holds list of all possible actions. */
    private static final List<Connect4Action> connect4Actions_ = new ArrayList<>();

    static {
        initActions();
    }

    /** Slot location to place piece. */
    private final int location_;

    private Connect4Action(int location) {
        location_ = location;
    }

    public Connect4Action copy() {
        return this;
    }

    private static void initActions() {
        for (int i = 0; i < Connect4State.WIDTH; i += 1) {
            connect4Actions_.add(new Connect4Action(i));
        }
    }

    /**
     * Returns the Connect 4 action representation of the slot location.
     * @param location
     *      slot location to place piece.
     *      value range from 0 to Connect4State.WIDTH - 1.
     * @return a Connect 4 action.
     */
    public static Connect4Action valueOf(int location) {
        return connect4Actions_.get(location);
    }

    public int getLocation() {
        return location_;
    }

    @Override
    public int hashCode() {
        return location_;
    }

    @Override
    public boolean equals(Object object) {
        if (!(object instanceof Connect4Action)) {
            return false;
        }
        Connect4Action connect4Action = (Connect4Action) object;
        return location_ == connect4Action.getLocation();
    }

    @Override
    public String toString() {
        return String.valueOf(location_ + 1);
    }
}

File: src/main/kotlin/com/castlefrog/agl/domains/connect4/Connect4Action.kt
package com.castlefrog.agl.domains.connect4

import com.castlefrog.agl.Action

/**
 * @param location slot location to place piece
 */
data class Connect4Action(val location: Int) : Action<Connect4Action> {

    override fun copy(): Connect4Action = this

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other?.javaClass != javaClass) return false
        return location == (other as Connect4Action).location
    }

    override fun hashCode(): Int {
        return location
    }

    override fun toString(): String {
        return (location + 1).toString()
    }

    companion object {
        /** Holds list of all possible actions. */
        private val actions = (0 until Connect4State.WIDTH).map(::Connect4Action)

        /**
         * Returns the Connect 4 action representation of the slot location.
         * @param location slot location to place piece.
         *                 value range from 0 to Connect4State.WIDTH - 1.
         * @return a Connect 4 action.
         */
        fun valueOf(location: Int): Connect4Action {
            return actions[location]
        }
    }
}
